api/commands/run.ts

```diff
@@ -61,134 +83,223 @@ export class Runner {
+          await this.executeAllActionsReadyForExecution();
+        })
       )
     ]);
   }
 
-  private async onActionExecutionComplete(executedAction: dataform.IExecutedAction) {
-    this.result.actions.push(executedAction);
-    await this.triggerChange();
-    await this.executeAllActionsReadyForExecution();
-  }
+  private async executeAction(action: dataform.IExecutionAction): Promise<void> {
+    if (action.tasks.length === 0) {
+      this.runResult.actions.push({
+        name: action.name,
+        status: dataform.ActionResult.ExecutionStatus.DISABLED,
+        tasks: []
+      });
+      await this.triggerChange();
+      return;
+    }
 
-  private async executeAction(action: dataform.IExecutionAction) {
-    const startTime = new Date().valueOf();
+    const timer = Timer.start();
+    const actionResult: dataform.IActionResult = {
+      name: action.name,
+      status: dataform.ActionResult.ExecutionStatus.RUNNING,
+      timing: timer.current(),
+      tasks: []
+    };
+    this.runResult.actions.push(actionResult);
+    await this.triggerChange();
 
-    const executedTasks: dataform.IExecutedTask[] = [];
-    let allSuccessful = true;
     for (const task of action.tasks) {
-      if (allSuccessful) {
-        const executedTask = await this.executeTask(task);
-        executedTasks.push(executedTask);
-        allSuccessful = allSuccessful && executedTask.ok;
+      if (this.cancelled) {
+        actionResult.tasks.push({
+          status: dataform.TaskResult.ExecutionStatus.CANCELLED
+        });
+        await this.triggerChange();
+      } else if (actionResult.status !== dataform.ActionResult.ExecutionStatus.RUNNING) {
+        actionResult.tasks.push({
+          status: dataform.TaskResult.ExecutionStatus.SKIPPED
+        });
+        await this.triggerChange();
+      } else {
+        const taskStatus = await this.executeTask(task, actionResult);
+        if (taskStatus === dataform.TaskResult.ExecutionStatus.FAILED) {
+          actionResult.status = dataform.ActionResult.ExecutionStatus.FAILED;
+        } else if (taskStatus === dataform.TaskResult.ExecutionStatus.CANCELLED) {
+          actionResult.status = dataform.ActionResult.ExecutionStatus.CANCELLED;
+        }
       }
     }
 
-    const endTime = new Date().valueOf();
-    const executionTime = endTime - startTime;
-    return {
-      name: action.name,
-      status: allSuccessful
-        ? executedTasks.length == 0
-          ? dataform.ActionExecutionStatus.DISABLED
-          : dataform.ActionExecutionStatus.SUCCESSFUL
-        : dataform.ActionExecutionStatus.FAILED,
-      tasks: executedTasks,
-      executionTime: Long.fromNumber(executionTime),
-      deprecatedOk: allSuccessful
-    };
+    if (actionResult.status === dataform.ActionResult.ExecutionStatus.RUNNING) {
+      actionResult.status = dataform.ActionResult.ExecutionStatus.SUCCESSFUL;
+    }
+    actionResult.timing = timer.end();
+    await this.triggerChange();
```

Comment:
This also looks like it will trigger twice for the final task if it executes successfully, once with just the task marked as successful, and once with the task and the surround action marked as successful.